lib/Sema/ConstraintSystem.cpp

```diff
@@ -3484,8 +3485,12 @@ void ConstraintSystem::bindOverloadType(
            "subscript always has one argument");
     // Parameter type is KeyPath<T, U> where `T` is a root type
     // and U is a leaf type (aka member type).
-    auto keyPathTy =
-        fnType->getParams()[0].getPlainType()->castTo<BoundGenericType>();
+    auto paramTy = fnType->getParams()[0].getPlainType();
+
+    if (auto *existential = paramTy->getAs<ExistentialType>())
+      paramTy = existential->getExistentialLayout().explicitSuperclass;
```

Comment:
do you want this to work?

```
protocol P: KeyPath<...> {...}

...

 subscript<V>(dynamicMember member: any P) -> V {
```

The more general way to get the superclass of an existential is existential->getSuperclass() instead of going through ExistentailLayout.

Also what if the existential type doesn't have a superclass? Is this code path reachable in that case? If so, you might crash on the next line because paramTy is null.

Reply:
Ok, makes sense. It's still simpler to write `paramTy = existential->getSuperclass()` though IMHO